### PR TITLE
Allow hostnames for fail2ban whitelist.

### LINKED ISSUE #2609: [Feature] Support hostnames for Fail2ban whitelist.
Since i have a dynamic ip address i can not whitelist it. 
On my hosts where i use fail2ban i can add my dyndnsname to avoid blocking me out.
Is there a chance to get this added?

### PR DESCRIPTION
**Pull request by Kraeutergarten**
Closes #2609

### CODE REVIEW

File: data/Dockerfiles/netfilter/server.py [hunk 1]
```diff
@@ -119,6 +127,19 @@ def ban(address):
   self_network = ipaddress.ip_network(address.decode('ascii'))
   if WHITELIST:
     for wl_key in WHITELIST:
+      if not is_ip_network(wl_key):
+        hostname = wl_key
+        try:
+          wl_key = socket.gethostbyname(hostname)
```
Comment (andryyy) [inline data/Dockerfiles/netfilter/server.py]: You know how long this takes to timeout if the name is invalid? :)

**Comment by andryyy** [comment]
How about blacklists?

**Comment by Kraeutergarten** [comment]
I don't see a usage for blacklists.

**Comment by andryyy** [comment]
Why?

**Comment by Kraeutergarten** [comment]
If i manually blacklist, i normally use the ip of the attacker or i block the network.

File: data/Dockerfiles/netfilter/server.py (hunk 1 above)
Comment (Kraeutergarten) [reply data/Dockerfiles/netfilter/server.py]: I don't think invalid names are a problem, the bigger problem is a slow responding dns.

Hm, what think you about a local cache? 
I looked into the Fail2ban source, they solve this issue with a cache ^^.

Maybe you have a suggestion?

**Edit:** I found a library where a timeout for DNS requests can be defined. Would this solve the problem? 

**PS:** Is there a reason for using python 2?

Comment (Kraeutergarten) [reply data/Dockerfiles/netfilter/server.py]: Timeout of 1 sec is ok?

**Comment by ntimo** [comment]
But not having this option for blacklists would be quite inconsistent :( And from a user perspective I would expect the black and whitelist to behave in the same way.

**Comment by Kraeutergarten** [comment]
I will try to add it.
How often should the whitelist/blacklist checked for changes? (1-5 minutes?)

**Comment by andryyy** [comment]
If i manually blacklist, i normally use the ip of the attacker or i block the network.

Hm, I see. I agree, but I also agree with @ntimo. Would be nice to not have that inconsistency. :) 
If you find time, it would be amazing to have this implemented. 

The whitelist is currently checked, whenever a rule was triggered. Blacklists apply on startup. Feel free to change that.

File: data/Dockerfiles/netfilter/server.py (hunk 1 above)
Comment (andryyy) [reply data/Dockerfiles/netfilter/server.py]: Oh, sorry, feel free to change it to whatever you think fits best!

Switching to Py3 would be great. No reason to stick with Py2. :) I think a timeout of 2-3s would be good.

**Comment by Kraeutergarten** [comment]
I would like to do blacklist resolving like the same. Whats do you think about it?

I really don't like how the "logging" was done. 
Hopefully you like my version?

**Comment by Kraeutergarten** [comment]
Now it's done. Maybe someone can help testing?

**Comment by andryyy** [comment]
Nice! I will try to test it asap.

**Comment by andryyy** [comment]
Yes, everything looks so much better now. Please let me use/test it for a few more hours.

**Comment by Kraeutergarten** [comment]
Of course, test as long as you want. 
If you have some suggestions, please let me know.

**Comment by ssuess** [comment]
So...if this was merged, I should be able to add a hostname (or partial hostname, or regex for hostnames) into the Blacklisted networks/hosts box in the Fail2ban config area?

**Pull request merged**